Allow up to 39 PVs attached to AWS node

jsafrane opened the pull request:
The devices are named /dev/xvdba - /dev/xvdcm, leaving /dev/sda - /dev/sdz for the
system.

AWS docs says:
- up to 40 devices are supported
- _recommended_ devices for EBS are /dev/sd[f-p]
- the same page it says that /dev/xvd[b-c][a-z] are _available_
  http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/device_naming.html#available-ec2-device-names

So, let's use all the _available_ devices, respecting limit of 40 devices.

Additional patches in this PR ensure that user gets some meaningful error when a node cannot attach more volumes to run a pod. When admin increases KUBE_MAX_PD_VOLS to 50. Only 39 pods are `Running`, the rest is `ContainerCreating` forever (or until some volumes are detached from the node) with "Too many EBS volumes attached to node" error:

```
$ kubectl describe pod
Events:
  FirstSeen     LastSeen        Count   From                                    SubobjectPath   Type            Reason          Message
  ---------     --------        -----   ----                                    -------------   --------        ------          -------
  8m            8m              1       {default-scheduler }                                    Normal          Scheduled       Successfully assigned pod-40 to ip-172-18-0-64.ec2.internal
  57s           57s             1       {kubelet ip-172-18-0-64.ec2.internal}                   Warning         FailedMount     Unable to mount volumes for pod "pod-40_default(1cf42d02-e9f0-11e5-94e6-0eed93379351)": Could not attach EBS Disk "vol-095e42aa": Too many EBS volumes attached to node XYZ.
  57s           57s             1       {kubelet ip-172-18-0-64.ec2.internal}                   Warning         FailedSync      Error syncing pod, skipping: Could not attach EBS Disk "vol-095e42aa": Too many EBS volumes attached to node XYZ.
```

@justinsb @kubernetes/sig-storage @kubernetes/rh-storage

jeremyeder commented:
@ekuric

ekuric commented:
I think there should  be an option with kubernetes to be able to attach more than Amazon recommended 40 EBS devices to Openshift node and start pods.

With 39 EBSs + one EBS for system, in best case there would be possible to start 39  pods per Openshift node ( some people can use additional devices for other purposes too, thus reducing number of devices available for pods ) 

Amazon recommends max 40 EBSs  ,as that can cause instability, but does not mean it will. (Attaching more than 40 volumes to a Linux instance is supported on a best effort basis only and is not guaranteed.). It is only from support point of view problematic to debug issues with many block devices attached

jsafrane commented:
I could rework the cloud provider to allow any number of devices, assuming that the scheduler never schedules pods with more than 39 PVs to a node. And scheduler can be overridden by `KUBE_MAX_PD_VOLS` env. variable if cluster admin really wants. This would apply to Kubernetes 1.2 and newer only.

@justinsb: thoughts?

justinsb commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -80,6 +80,11 @@ const MaxReadThenCreateRetries = 30
+// Amazon recommends having no more that 40 volumes attached to an instance,
+// and at least one of those is for the system root volume.
+// See http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/volume_limits.html#linux-specific-volume-limits
+const DefaultMaxEBSVolumes = 39
```
Yes, I think I do like the idea of having the scheduler restrict this.  I don't mind if you want to keep this here and raise a warning if we exceed 39 volumes... I expect that will save us a lot of time debugging :-)

justinsb commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -902,9 +907,17 @@ type mountDevice string
 	devices := []mountDevice{}
-	for c := 'f'; c <= 'p'; c++ {
-		devices = append(devices, mountDevice(fmt.Sprintf("%c", c)))
+	for i, first, second := 0, 'b', 'a'; i < DefaultMaxEBSVolumes; i++ {
```
This might be clearer written as two loops (over first, then over second, with a separate check for # of volumes.  But it's up to you!

jsafrane replied:
ok

justinsb commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -1014,7 +1027,7 @@ func (self *awsInstance) getMountDevice(volumeID string, assign bool) (assigned
 			name = name[8:]
 		}
-		if len(name) != 1 {
+		if len(name) < 1 {
```
The intention was to catch 'dev' still in there.  How about we check for len(name) < 1 || len(name) > 2

jsafrane replied:
ok

justinsb commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -1051,7 +1064,7 @@ func (self *awsInstance) getMountDevice(volumeID string, assign bool) (assigned
 	if chosen == "" {
 		glog.Warningf("Could not assign a mount device (all in use?).  mappings=%v, valid=%v", deviceMappings, valid)
-		return "", false, nil
+		return "", false, fmt.Errorf("Too many EBS volumes attached to node %s.", instanceName)
```
Ah - OK.   I had meant for this not to be an error, but for the caller to handle.  But your way is better (and I forgot to have the caller handle, so my way plainly is too hard!)

justinsb commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -989,7 +1002,7 @@ func (self *awsInstance) describeInstance() (*ec2.Instance, error) {
 // If the volume is already assigned, this will return the existing mountDevice with alreadyAttached=true.
 // Otherwise the mountDevice is assigned by finding the first available mountDevice, and it is returned with alreadyAttached=false.
-func (self *awsInstance) getMountDevice(volumeID string, assign bool) (assigned mountDevice, alreadyAttached bool, err error) {
+func (self *awsInstance) getMountDevice(volumeID string, assign bool, instanceName string) (assigned mountDevice, alreadyAttached bool, err error) {
```
You don't need to pass in instanceName, because this is a member function of awsInstance.  So you can just get it by `self.nodeName`

jsafrane replied:
good point

justinsb commented:
Generally this looks good to me as-is.  There's definitely a correctness issue going on.  I'm going to try testing this to make sure that these "two letter" volume attachments do work, but I think we should merge this.  Probably even cherry-pick to 1.2.

As for attaching > 39 volumes, I think maybe we can have that be a separate PR.  I don't know if it is a good idea (we'll have all sorts of random attachment errors it sounds like), but if it is important to OpenShift then I'm open to it.

jeremyeder commented:
@justinsb the defaults being safe, supported from the cloud provider is mandatory, and seems like we have that now whether enforced by that const or by the scheduler.

As far as whether this belongs in 1.2, this is at least internally on our "blocker" list, so either we cherry pick it ourselves (ugh) or we get it for free with 1.2 :-)

My team would like an override if possible (it can be wrapped in warnings about non-support etc...), because we would like to do exploratory work on that limit.

@jsafrane and so we would just throw ```KUBE_MAX_PD_VOLS=200``` in ```/etc/sysconfig/atomic-openshift-master``` and restart?  Perfect, if that's what you meant!

justinsb commented:
No objection to an override flag if you can rush it in.  I'd love to get this into the release though, and I think we might already be too late for it.

justinsb commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -899,13 +905,38 @@ type awsInstanceType struct {
-		devices = append(devices, mountDevice(fmt.Sprintf("%c", c)))
+	i := 0
+	count := getMaxVols(DefaultMaxEBSVolumes)
+	for first := 'a'; i < count; first++ {
```
I think first should start at 'b', right?

justinsb commented:
One nit (we should start the first loop at b I think) but otherwise LGTM

jsafrane commented:
Fixed all review remarks.

There is no override yet, "39" is still hardcoded. I'll try to prepare something tomorrow, probably by removing whole `getEBSMountDevices` - kubelet will attach all devices that were scheduled to the node. It will be up to the scheduler to restrict it to "39" or any other limit that the admin configures with `KUBE_MAX_PD_VOLS` env. variable.

jsafrane commented:
@justinsb sorry, I pushed older version (with ugly copy-paste from scheduler), please review now.

justinsb commented:
Awesome - LGTM.  If you want to include the override using the env var, that looked fine to me as well (it's important to OpenShift, right?)

jsafrane committed:
Allow 39 atached EBS devices.

AWS has soft support limit for 40 attached EBS devices. Assuming there is just
one root device, use the rest for persistent volumes.

The devices will have name /dev/xvdba - /dev/xvdcm, leaving /dev/sda - /dev/sdz
to the system.

Also, add better error handling and propagate error
"Too many EBS volumes attached to node XYZ" to a pod.

jsafrane commented:
Squashed.

> If you want to include the override using the env var

It was ugly when I saw it :-). Copy-paste instead of proper solution.

> it's important to OpenShift, right

I think it's important for experiments now, 40 volumes attached to a node is not too much. We need to test it with more. @jeremyeder can say more.

justinsb commented:
Thanks for tracking down the longer list of volume names, BTW.  I'll re-add LGTM if mergebot removes it.

jeremyeder commented:
Thanks guys -- don't want to stop this merge.  I've filed a different PR for the override:  https://github.com/kubernetes/kubernetes/issues/22994

eparis commented:
I'm using this PR to debug the cherrypick queue. Please ignore it. It has definitely been merged to release 1.2 and the cherrypick-* flags no longer matter.
